src/linux_android.rs

```diff
@@ -1,19 +1,80 @@
 //! Implementation for Linux / Android without `/dev/urandom` fallback
-use crate::{util_libc, Error};
+use crate::{util_unix::sys_fill_exact, Error};
 use core::mem::MaybeUninit;
 
 pub fn getrandom_inner(dest: &mut [MaybeUninit<u8>]) -> Result<(), Error> {
-    util_libc::sys_fill_exact(dest, getrandom_syscall)
+    sys_fill_exact(dest, getrandom_syscall)
 }
 
+// The value of `EINTR` is not architecture-specific. It is checked against
+// `libc::EINTR` by linux_android_with_fallback.rs.
+pub const EINTR: i32 = 4;
+
 // Also used by linux_android_with_fallback to check if the syscall is available.
-pub fn getrandom_syscall(buf: &mut [MaybeUninit<u8>]) -> libc::ssize_t {
-    unsafe {
-        libc::syscall(
-            libc::SYS_getrandom,
-            buf.as_mut_ptr().cast::<core::ffi::c_void>(),
-            buf.len(),
-            0,
-        ) as libc::ssize_t
+cfg_if! {
+    // TODO: Expand inilne assembly to other architectures to avoid depending
+    // on libc on Linux.
+    if #[cfg(all(target_arch = "x86_64", target_pointer_width = "64"))] {
+        type Word = u64;
+        type IWord = i64;
+
+        // TODO(MSRV(1.78 feature(target_abi))): Enable this and remove `target_pointer_width`
+        // restriction above.
+        //
+        // #[cfg(target_abi = "x32")]
+        // const __X32_SYSCALL_BIT: Word = 0x40000000;
+        //
+        // #[cfg(target_abi = "x832")]
+        // #[allow(non_upper_case_globals)]
+        // pub const SYS_getrandom: IWord = 318 | __X32_SYSCALL_BIT;
```

Comment:
It isn't clear to me that we actually need to use `__X32_SYSCALL_BIT` for `x86_64-unknown-linux-gnux32` and other x32 targets. It seems like we could just use the 64-bit version of the syscall instead (i.e. same syscall number). But I am not so sure and it's hard for me to test x32.